Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp

```diff
@@ -167,6 +168,22 @@ static void setCGBlendMode(CGContextRef context, CompositeOperator op, BlendMode
     CGContextSetBlendMode(context, selectCGBlendMode(op, blendMode));
 }
 
+static void setCGContextPath(CGContextRef context, const Path& path)
+{
+    CGContextBeginPath(context);
+    addToCGContextPath(context, path);
+}
+
+static void drawPathWithCGContext(CGContextRef context, CGPathDrawingMode drawingMode, const Path& path)
+{
+#if HAVE(CG_CONTEXT_DRAW_PATH_DIRECT)
+    CGContextDrawPathDirect(context, drawingMode, path.platformPath(), nullptr);
+#else
+    setCGContextPath(context, path);
+    CGContextDrawPath(context, drawingMode);
+#endif
```

Comment:
I think we should remove `HAVE(CG_CONTEXT_DRAW_PATH_DIRECT)` and the `#else` part since HAVE_CG_CONTEXT_DRAW_PATH_DIRECT is true for all `PLATFORM(COCOA)`. Maybe in another patch is fine. But this means drawPath() does not benefit for this change.

Reply:
> But this means drawPath() does not benefit for this change.

Yeah, this change is not a progression with drawPath for our current benchmarks.

Reply:
To add, as much as I could gather, this is due to when we add segments via CGContextAdd*, the segments are transformed with CGContext CTM. This seems to be slow when done per call.

When creating an untransformed path and then drawing it directly, the draw call can do the transform book keeping once and then iterate over the path segments, transforming as it goes. Perhaps also some paths don't need the transform, rather the transform can be an argument.